### ISSUE TITLE
Paramspec used within union reports errors that paramspec is missing

### ISSUE THREAD

**Issue by Dr-Irv** [microsoft/pyright#6796]
**Describe the bug**
If you have a function that accepts a union based on a callable using `ParamSpec` or a `Callable` that doesn't, pyright reports an error in calls to that function

**Code or Screenshots**

```python
from typing import Callable, Concatenate
from typing_extensions import ParamSpec

P = ParamSpec("P")


def func1(x: int):
    print(x)


def func2(x: int, /, y: str):
    print(x)


def accept_func(
    f: Callable[Concatenate[int, P], None] | tuple[int, Callable[..., None]],
    *args: P.args,
    **kwargs: P.kwargs,
) -> None:
    if isinstance(f, tuple):
        if args:
            f[1](f[0], args)
        else:
            f[1](f[0])
    else:
        f(5)


accept_func(func1)
accept_func((2, func1))
accept_func((3, func2), "abc")

```

pyright reports:
```
 pspec.py:26:9 - error: Arguments for ParamSpec "P@accept_func" are missing (reportGeneralTypeIssues)
 pspec.py:29:1 - error: Arguments for ParamSpec "P@accept_func" are missing (reportGeneralTypeIssues)
 pspec.py:30:25 - error: Arguments for ParamSpec "P@accept_func" are missing (reportGeneralTypeIssues)
```

mypy 1.7.1 accepts this.  Note - we have a PR for pandas-stubs that is trying to use this pattern.

**VS Code extension or command-line**
- Both - version 1.1.342 on command line
- python 3.11.3

**Comment by erictraut** [comment]
Upon closer inspection...

The first of the three errors emitted by pyright is correct, so this isn't a pyright bug. This appears to be a false negative in the latest version of mypy. 

For the latter two errors, it's not clear to me whether this construct should be allowed. You're trying to exploit a gray area of the typing spec. By a strict reading of PEP 612, I think this shouldn't be allowed. The only way it would be allowable is if an unsolved ParamSpec is assumed to be equivalent to `...` — that is, the signature should accept any arguments. That would mean all of the following would be allowed:
* `accept_func((2, func1))`
* `accept_func((2, func1), "")
* `accept_func((2, func1), foo="", bar=1)

There would be effectively no type checking for the argument list (beyond the first tuple argument), which would be really unsafe and counter to the intended use of `ParamSpec`. Is that really the behavior you want here?

I suspect that the latter two errors are also a bug (a false negative) in mypy's current ParamSpec implementation — probably just an oversight in its current implementation. You should definitely check with them before exploiting this behavior.

I also recommend posting to the [Python/Typing discourse forum](https://discuss.python.org/c/typing/32) to get opinions from the typing community about whether the above should be allowed. 

I'm going to close this as "working as designed" for now because I think that's the right answer from a type standpoint. If you want to pursue this approach and you get confirmation from the mypy maintainers and a general agreement from the typing community that this usage should be supported, then I'll make the change in pyright.

In the meantime, I recommend using an overload rather than a union.

Code sample in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCCKcANFAMICGANtZQEbUCmZ5YKAxpTEyty1DAA3JiGphKAEwCwAKFCRYiVGgD6TAB48UAZyTsdmbLnwAFSiEoQAygiYc5c01AC8Uc5Zt2OACgBEpn4AlI6ykkzAUMAArpwAjD4aAFyYKDBBSXJQ2VAIIKgwiSGycuGRMZwATIkpBWQA9GRwKTowIBlZOXkFRaEAAsKi4lKlEVCUHBxMCDCqFb6d2cApMNEIzADadRQ0dIxMGwB0x2QAcuxMALqXJItQAFQWaDopRKR39-cA1gDuTy%2BEYi3WRBKAAWgAfFBzigmJlZDkoMdDv1BmIJDIwmMJlMZnNYgsETlljtaAxNmxOPw%2BDwtmkyKYbtCLjcPv8UqZDv9gYjPr92e5DvyQM9gaDIczYfDEcjQmVxpNprN5j47iSqGT9htKVxtPw6TAGUyYVcoAAfWBrTbbDV7TbIs4s1lE7KPEUAzncj7fP7ujlC32iuTiqEm6U5JCRJB6XQwSicJg%2BYBkVbrJgdF2IzCRdl3LPEjZxS5JjYABiZ-2K%2BeyTGoOjheerwELxeb5arOVr9fD%2BeAPgArGQ3aKHj7K6E5DilfjOEmCXFilO8SqfJUyPMF4vFcuCT4fABmdcEypBMh%2BBgcYJyIA)

```python
from typing import Any, Callable, Concatenate, overload
from typing_extensions import ParamSpec

P = ParamSpec("P")

def func1(x: int):
    print(x)

def func2(x: int, /, y: str):
    print(x)

@overload
def accept_func(
    f: tuple[int, Callable[..., None]],
    *args: Any,
    **kwargs: Any,
) -> None:
    ...

@overload
def accept_func(
    f: Callable[Concatenate[int, P], None],
    *args: P.args,
    **kwargs: P.kwargs,
) -> None:
    ...

def accept_func(
    f: Callable[Concatenate[int, P], None] | tuple[int, Callable[..., None]],
    *args: P.args,
    **kwargs: P.kwargs,
) -> None:
    if isinstance(f, tuple):
        if args:
            f[1](f[0], args)
        else:
            f[1](f[0])
    else:
        f(5, *args, **kwargs)

accept_func(func1)
accept_func((2, func1))
accept_func((3, func2), "abc")
```

**Comment by Dr-Irv** [comment]
> There would be effectively no type checking for the argument list (beyond the first tuple argument), which would be really unsafe and counter to the intended use of `ParamSpec`. Is that really the behavior you want here?

Yes. We are trying to support `DataFrame.pipe()` (see https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.pipe.html), which allows any function to be passed. 

Signature:  `DataFrame.pipe(func, *args, **kwargs)`

`func` can be:
Function to apply to the Series/DataFrame. args, and kwargs are passed into func. Alternatively a (callable, data_keyword) tuple where data_keyword is a string indicating the keyword of callable that expects the Series/DataFrame.

So we know there won't be type checking for users on the args/kwargs that are passed, but we need to support arbitrary function signatures.

I agree there is a bunch of ambiguity here, because the function signatures that are passed are arbitrary, which means you lose the type checking anyway with the use of `*args` and `**kwargs` .

**Comment by Dr-Irv** [comment]
So I tried your overload idea, but this time with a class, which more closely mimics what we are trying to do with pandas:

```python
from typing import Callable, Concatenate, overload, Any
from typing_extensions import ParamSpec, Self

P = ParamSpec("P")


class Foo:
    @overload
    def accept_func(
        self,
        f: Callable[Concatenate[Self, P], None],
        *args: P.args,
        **kwargs: P.kwargs,
    ) -> None:
        ...

    @overload
    def accept_func(
        self,
        f: tuple[Callable[..., None], str],
        *args: Any,
        **kwargs: Any,
    ) -> None:
        ...

    def accept_func(
        self,
        f,
        *args,
        **kwargs,
    ) -> None:
        if isinstance(f, tuple):
            f[0](self, *args, **kwargs)
        else:
            f(self, *args, **kwargs)


def func1(fo: Foo, x: int):
    print("func1", x)


def func2(fo: Foo, x: int, /, y: str):
    print("func2", x, "y", y)


def func3(fo: Foo, x: int, /, y: str, *, z: tuple[int, int]):
    print("func3", x, "y", y, "z", z)


foo = Foo()
foo.accept_func(func1, 1)
foo.accept_func(func2, 3, "abc")   # error reported here
foo.accept_func(func3, 6, "def", z=(10, 20))   # error reported here

foo.accept_func((func1, "f1"), 10)
foo.accept_func((func2, "f2"), 30, "abc")
foo.accept_func((func3, "f3"), 40, "def", z=(30, 40))
```

Now pyright reports:
```text
pspecmultargs.py:52:1 - error: No overloads for "accept_func" match the provided arguments (reportGeneralTypeIssues)
pspecmultargs.py:53:1 - error: No overloads for "accept_func" match the provided arguments (reportGeneralTypeIssues)
```

mypy accepts this.  The error is reported on the annotated lines.

**Comment by erictraut** [comment]
Reopening issue. Yes, the errors in the above code sample are false positives. I'm surprised no one has reported this one previously.

I've introduced a fix and updated our unit tests to include a variant of your code sample.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.343, which I just published. It will also be included in a future release of pylance.
